framework/components/dockercompose/chip_ingress_set/protos.go

```diff
@@ -422,74 +452,126 @@ type schemaStatus struct {
-// it keeps doing this until all protos are registered or it fails to register any more protos
-func registerAllWithTopologicalSortingByTrial(
+// registerAllWithTopologicalSorting registers protos in dependency order using topological sorting
+func registerAllWithTopologicalSorting(
 	schemaRegistryURL string,
 	protoMap map[string]string, // path -> proto source
 	subjectMap map[string]string, // path -> subject
 ) error {
-	framework.L.Info().Msgf("🔄 registering %d protobuf schemas", len(protoMap))
+	framework.L.Info().Msgf("Registering %d protobuf schemas", len(protoMap))
+
+	// Build dependency graph and sort topologically
+	dependencies, depErr := buildDependencyGraph(protoMap)
+	if depErr != nil {
+		return errors.Wrap(depErr, "failed to build dependency graph")
+	}
+
+	sortedFiles, sortErr := topologicalSort(dependencies)
+	if sortErr != nil {
+		return errors.Wrap(sortErr, "failed to sort files topologically")
+	}
+
+	framework.L.Info().Msgf("Registration order (topologically sorted): %v", sortedFiles)
+
 	schemas := map[string]*schemaStatus{}
 	for path, src := range protoMap {
 		schemas[path] = &schemaStatus{Source: src}
 	}
 
-	refs := []map[string]any{}
-
-	for {
-		progress := false
-		failures := []string{}
+	// Register files in topological order
+	for _, path := range sortedFiles {
+		schema, exists := schemas[path]
+		if !exists {
+			framework.L.Warn().Msgf("File %s not found in schemas map", path)
+			continue
+		}
 
-		for path, schema := range schemas {
-			if schema.Registered {
-				continue
-			}
+		if schema.Registered {
+			continue
+		}
 
-			subject, ok := subjectMap[path]
-			if !ok {
-				failures = append(failures, fmt.Sprintf("%s: no subject found", path))
-				continue
-			}
+		subject, ok := subjectMap[path]
+		if !ok {
+			return fmt.Errorf("no subject found for %s", path)
+		}
 
-			singleProtoFailures := []error{}
-			framework.L.Debug().Msgf("🔄 registering %s as %s", path, subject)
-			_, registerErr := registerSingleProto(schemaRegistryURL, subject, schema.Source, refs)
-			if registerErr != nil {
-				failures = append(failures, fmt.Sprintf("%s: %v", path, registerErr))
-				singleProtoFailures = append(singleProtoFailures, registerErr)
-				continue
+		// Build references only for files that have dependencies
+		var fileRefs []map[string]any
+		if deps, hasDeps := dependencies[path]; hasDeps && len(deps) > 0 {
+			for _, dep := range deps {
+				if depSubject, depExists := subjectMap[dep]; depExists {
+					// The schema registry expects import names without the 'workflows/' prefix
+					// So if the import is "workflows/v1/metadata.proto", the name should be "v1/metadata.proto"
+					// And if the import is "workflows/v2/cre_info.proto", the name should be "v2/cre_info.proto"
+					importName := dep
+					if strings.HasPrefix(dep, "workflows/") {
```

Comment:
this will only work with current folder structure. That's why I was removing that [prefix dynamically](https://github.com/smartcontractkit/chainlink-testing-framework/pull/2036/files#r2270553625). I'd rather not hardcode that `workflows/` removal and keep it dynamic.